common/controllers/fps-controller.cpp

```diff
@@ -20,76 +21,153 @@ namespace tec {
 		int strafe = 0;
 
-		if (this->KEY_W_DOWN && this->KEY_W_FIRST) {
+		// if (CLIENT_KEY_EVENT || SERVER_CLIENT_COMMAND)
+		if ((this->KEY_W_DOWN && this->KEY_W_FIRST) || this->forward) {
 			forward = -1;
+			this->forward = true;
 		}
-		else if (this->KEY_S_DOWN) {
+		else if (this->KEY_S_DOWN || this->backward) {
 			forward = 1;
+			this->backward = true;
 		}
-		if (this->KEY_A_DOWN && this->KEY_A_FIRST) {
+		if ((this->KEY_A_DOWN && this->KEY_A_FIRST) || this->left_strafe) {
 			strafe = -1;
+			this->left_strafe = true;
 		}
-		else if (this->KEY_D_DOWN) {
+		else if (this->KEY_D_DOWN || this->right_strafe) {
 			strafe = 1;
+			this->right_strafe = true;
 		}
 
-		glm::quat orientation;
 		if (state.orientations.find(entity_id) != state.orientations.end()) {
-			orientation = state.orientations.at(entity_id).value;
+			const_cast<GameState&>(state).orientations[entity_id] = this->orientation;
 		}
 
-		const_cast<GameState&>(state).velocities[entity_id].linear = glm::vec4(orientation * glm::vec3(7.0 * strafe, 0.0, 7.0 * forward), 1.0);
+		const_cast<GameState&>(state).velocities[entity_id].linear = glm::vec4(this->orientation.value * glm::vec3(3.0 * strafe, 0.0, 7.5 * forward), 1.0);
+	}
+
+	proto::ClientCommands FPSController::GetClientCommands() {
+		proto::ClientCommands proto_client_commands;
+		proto_client_commands.set_id(this->entity_id);
+		proto::MovementCommand* movement_command = nullptr;
+		if (this->forward) {
+			if (!movement_command) {
+				movement_command = proto_client_commands.mutable_movement();
+				movement_command->set_backward(false);
+				movement_command->set_leftstrafe(false);
+				movement_command->set_rightstrafe(false);
+			}
+			movement_command->set_forward(true);
+		}
+		else if (this->backward) {
+			if (!movement_command) {
+				movement_command = proto_client_commands.mutable_movement();
+				movement_command->set_forward(false);
+				movement_command->set_leftstrafe(false);
+				movement_command->set_rightstrafe(false);
+			}
+			movement_command->set_backward(true);
+		}
+		if (this->left_strafe) {
+			if (!movement_command) {
+				movement_command = proto_client_commands.mutable_movement();
+				movement_command->set_forward(false);
+				movement_command->set_backward(false);
+				movement_command->set_rightstrafe(false);
+			}
+			movement_command->set_leftstrafe(true);
+		}
+		else if (this->right_strafe) {
+			if (!movement_command) {
+				movement_command = proto_client_commands.mutable_movement();
+				movement_command->set_forward(false);
+				movement_command->set_backward(false);
+				movement_command->set_leftstrafe(false);
+			}
+			movement_command->set_rightstrafe(true);
+		}
+		auto orientation = proto_client_commands.mutable_orientation();
+		orientation->set_x(this->orientation.value.x);
+		orientation->set_y(this->orientation.value.y);
+		orientation->set_z(this->orientation.value.z);
+		orientation->set_w(this->orientation.value.w);
+		return std::move(proto_client_commands);
+	}
```

Comment:
Might be shorter and easier in this function to initialise the movement_command at the beginning and then just set the properties?